lib/translator/read.cc

```diff
@@ -0,0 +1,187 @@
+#include "read.h"
+
+#include "absl/types/span.h"
+
+#include "common.h"
+
+namespace translator {
+
+namespace {  // anonymous namespace for helper functions
+
+// Section 5.3
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+StatusCode BuildPrinfo(uint8_t rdprotect, uint8_t& prinfo) {
+  bool pract;     // Protection Information Action 1 bit
+  uint8_t prchk;  // Protection Information Check 3 bits
+  prinfo = 0;     // Protection Information field 4 bits
+
+  switch (rdprotect) {  // 3 bits
+    case 0b0:
+      pract = 0b1;
+      prchk = 0b111;
+      break;
+    case 0b001:
+    case 0b101:
+      pract = 0b0;
+      prchk = 0b111;
+      break;
+    case 0b010:
+      pract = 0b0;
+      prchk = 0b011;
+      break;
+    case 0b011:
+      pract = 0b0;
+      prchk = 0b000;
+      break;
+    case 0b100:
+      pract = 0b0;
+      prchk = 0b100;
+      break;
+    default:
+      // Should result in SCSI command termination with status: CHECK CONDITION,
+      //  sense key: ILLEGAL REQUEST, additional sense code: LLEGAL FIELD IN CDB
+      DebugLog("RDPROTECT with value %d has no translation to PRINFO",
+               rdprotect);
+      return StatusCode::kInvalidInput;
+  }
+
+  prinfo |= prchk;
+  prinfo |= (pract << 3);
+
+  return StatusCode::kSuccess;
+}
+
+void SetLbaTags(uint32_t eilbrt, uint16_t elbat, uint16_t elbatm,
+                nvme::GenericQueueEntryCmd& nvme_cmd) {
+  // cdw14 expected initial logical block reference
+  nvme_cmd.cdw[4] = eilbrt;
+  // cdw15 bits 15:0 expected logical block application tag
+  nvme_cmd.cdw[5] |= elbat;
+  // cdw15 bits 31:16 expected logical block application tag mask
+  nvme_cmd.cdw[5] |= (elbatm << 16);
+}
+
+// Translates fields common to all Read commands
+// Named Legacy because it is called directly by Read6, an obsolete command
+// lacking fields common to other Read commands
+StatusCode LegacyRead(uint32_t lba, uint16_t transfer_length,
+                      nvme::GenericQueueEntryCmd& nvme_cmd,
+                      Allocation& allocation) {
+  StatusCode status_code = allocation.SetPages(1, 1);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  nvme_cmd = nvme::GenericQueueEntryCmd{
+      .opc = static_cast<uint8_t>(nvme::NvmOpcode::kRead),
+      .psdt = 0b00  // PRPs are used for data transfer
+  };
+
+  nvme_cmd.mptr = allocation.mdata_addr;
+  nvme_cmd.dptr.prp.prp1 = allocation.data_addr;
+  nvme_cmd.cdw[0] = lba;              // cdw10 Starting lba bits 31:00
+  nvme_cmd.cdw[2] = transfer_length;  // cdw12 nlb bits 15:00
```

Comment:
In nvme, the transfer length is a 0-based number of blocks; I.e. a value of 0 here means to read 1 block.  I think you'll need to include an adjustment factor here.

Reply:
thanks for the catch!